### PR TITLE
[OSSALifetimeCompletion] Handle unavailable blocks in dead-end regions.

### PR DESCRIPTION
**Pull request by nate-chandler**
Not every block in a region which begins with the non-lifetime-ending boundary of a value and ending with `unreachable`-terminated blocks has the value available.  If the `unreachable`-terminated blocks in this boundary are not available, it is incorrect to insert destroys of the value in them: it is an overconsume on some paths.  Previously, however, destroys were simply being inserted at the `unreachable`.

Here, this is fixed by finding the boundary of availability within that region and inserting destroys before the terminators of the blocks on that boundary.

rdar://116255254

### CODE REVIEW

File: test/SILOptimizer/mem2reg_lifetime.sil [hunk 1]
```diff
@@ -1858,8 +1858,7 @@ right:
 // CHECK:       {{bb[^,]+}}([[INSTANCE:%[^,]+]] : @owned $C):
 // CHECK:         cond_br undef, [[BB1:bb[0-9]+]], [[BB4:bb[0-9]+]]
 // CHECK:       [[BB1]]:
-// CHECK:         [[COPY:%[^,]+]] = copy_value [[INSTANCE]]
```
Comment (atrick) [inline test/SILOptimizer/mem2reg_lifetime.sil:1861]: I'm curious why this change eliminated a copy in mem2reg.

File: include/swift/SILOptimizer/Utils/CanonicalizeOSSALifetime.h [hunk 2]
```diff
@@ -409,6 +421,10 @@ class CanonicalizeOSSALifetime final {
   void recordConsumingUser(SILInstruction *user) {
     consumingBlocks.insert(user->getParent());
   }
+  void recordUnreachableLifetimeEnd(SILInstruction *user) {
```
Comment (atrick) [inline include/swift/SILOptimizer/Utils/CanonicalizeOSSALifetime.h:424]: At the definition of `consumingBlocks` and `recordConsumingUser` we should comment that these are either the points where the original value was consumed or was available at an unreachable instruction.

**Review by atrick — APPROVED** [review]
The forward data flow looks great. Thanks. This was a big missing OSSA piece.

File: test/SILOptimizer/mem2reg_lifetime.sil (hunk 1 above)
Comment (nate-chandler) [reply test/SILOptimizer/mem2reg_lifetime.sil:1861]: This copy was incorrectly introduced in https://github.com/apple/swift/pull/68608/files#diff-8bd51ca385c836620fb33d8b5480d0ab38024c0ba38f83aa599f0a5148c48e7fR1861 .  It was introduced because the unreachable instruction was added to liveness although the value is not available at it.  This is fixed here by not adding unavailable terminators to liveness.

File: include/swift/SILOptimizer/Utils/CanonicalizeOSSALifetime.h (hunk 2 above)
Comment (nate-chandler) [reply include/swift/SILOptimizer/Utils/CanonicalizeOSSALifetime.h:424]: Fixed.

**Comment by nate-chandler** [comment]
@swift-ci please test

**Comment by meg-gupta** [comment]
thank you!
